registry/storage/catalog.go

```diff
@@ -29,32 +29,33 @@ func (reg *registry) Repositories(ctx context.Context, repos []string, last stri
 
 	err = reg.blobStore.driver.Walk(ctx, root, func(fileInfo driver.FileInfo) error {
 		err := handleRepository(fileInfo, root, last, func(repoPath string) error {
-			foundRepos = append(foundRepos, repoPath)
+			repos[foundRepos] = repoPath
+			foundRepos += 1
 			return nil
 		})
 		if err != nil {
 			return err
 		}
 
-		// if we've filled our array, no need to walk any further
-		if len(foundRepos) == len(repos) {
+		// if we've filled our slice, no need to walk any further
+		if foundRepos == len(repos) {
 			finishedWalk = true
 			return driver.ErrSkipDir
 		}
 
 		return nil
 	})
 
-	n = copy(repos, foundRepos)
-
 	if err != nil {
-		return n, err
-	} else if !finishedWalk {
-		// We didn't fill buffer. No more records are available.
-		return n, io.EOF
+		return foundRepos, err
 	}
 
-	return n, err
+	if !finishedWalk {
+		// We didn't fill buffer. No more records are available.
+		return foundRepos, io.EOF
+	} else {
```

Comment:
I'm not sure what you mean, we need to return the number of found repos in either case.

Reply:
Sorry, unless I'm misreading that code I'm suggesting to change this
```Go
if !finishedWalk {
	// We didn't fill buffer. No more records are available.
	return foundRepos, io.EOF
} else {
	return foundRepos, err
}
```

to this
```Go
if !finishedWalk {
	// We didn't fill buffer. No more records are available.
	return foundRepos, io.EOF
}
// we know the walk is finished so let's return
return foundRepos, err
```

Reply:
Oh right, a style thing :)

When both branches are returns, I like to keep them together. Its not an escape block because something's wrong, that whole if statement is "We've done our job, what do we tell our caller?"

I guess filling the buffer is really the escape block where we haven't finished, so have refactored. Wdyt?